### PR TITLE
Merged PR 9371: Revive support for globalization and localization in Blazor WASM

### LINKED ISSUE #22975: Figure out how timezones work in browser-wasm
* How does the tz database get referenced?
* Can you disable it to save bytes?

### LINKED ISSUE #23260: Revive Blazor Wasm globalization \ localization tests
Currently blocked on https://github.com/dotnet/runtime/issues/38124

### LINKED ISSUE #24174: Enable ICU in blazor
Since the blazor platform code calls directly into the runtime rather than using the normal initialization code it doesn't have the logic to load the icudt.dat file into the runtime that is required to enable icu.  It will need to copy `icudt.dat` from the runtime pack and insert it into memory using `offset = this.mono_wasm_load_bytes_into_heap (bytes)` and `this.mono_wasm_load_icu_data (offset)` and only do the logic in https://github.com/dotnet/aspnetcore/pull/24163 if that file doesn't exist

### PR DESCRIPTION
**Pull request by pranavkm**
Manually porting preview8 changes to master

Revive support for globalization and localization in Blazor WASM

* Load icu and timezone data files
* Unskip tests

Fixes https://github.com/dotnet/aspnetcore/issues/24174
Fixes https://github.com/dotnet/aspnetcore/issues/22975
Fixes https://github.com/dotnet/aspnetcore/issues/23260

### CODE REVIEW

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts [hunk 1]
```diff
@@ -239,14 +239,26 @@ function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourceLoade
     /* hash */ resourceLoader.bootConfig.resources.runtime[dotnetWasmResourceName],
     /* type */ 'dotnetwasm');
 
-  const dotnetTimeZoneResourceName = 'dotnet.timezones.dat';
+  const dotnetTimeZoneResourceName = 'dotnet.timezones.blat';
```
Comment (captainsafia) [inline src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:242]: Is this a typo or a legit file format?

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts [hunk 2]
```diff
@@ -358,7 +374,11 @@ function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourceLoade
     resourceLoader.purgeUnusedCacheEntriesAsync(); // Don't await - it's fine to run in background
 
     MONO.mono_wasm_setenv("MONO_URI_DOTNETRELATIVEORABSOLUTE", "true");
-    MONO.mono_wasm_setenv("DOTNET_SYSTEM_GLOBALIZATION_INVARIANT", "1");
+    let timeZone = "UTC";
+    try {
+      timeZone = Intl.DateTimeFormat().resolvedOptions().timeZone;
+    } catch { }
```
Comment (captainsafia) [inline src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:380]: Why is it OK to ignore errors thrown here?

File: src/Components/WebAssembly/Sdk/integrationtests/WasmBuildIntegrationTest.cs [hunk 3]
```diff
@@ -192,10 +193,10 @@ public async Task Build_WithBlazorEnableTimeZoneSupportDisabled_DoesNotCopyTimeZ
 
             var runtime = bootJsonData.resources.runtime.Keys;
             Assert.Contains("dotnet.wasm", runtime);
-            Assert.DoesNotContain("dotnet.timezones.dat", runtime);
+            Assert.DoesNotContain("dotnet.timezones.blat", runtime);
```
Comment (captainsafia) [inline src/Components/WebAssembly/Sdk/integrationtests/WasmBuildIntegrationTest.cs:196]: Ah -- OK. So the `blat` file extension is intentional...

File: src/Components/WebAssembly/WebAssembly/src/Hosting/SatelliteResourcesLoader.cs [hunk 4]
```diff
@@ -55,7 +57,8 @@ public virtual async ValueTask LoadCurrentCultureResourcesAsync()
 
             for (var i = 0; i < assemblies.Length; i++)
             {
-                Assembly.Load((byte[])assemblies[i]);
+                using var stream = new MemoryStream((byte[])assemblies[i]);
```
Comment (captainsafia) [inline src/Components/WebAssembly/WebAssembly/src/Hosting/SatelliteResourcesLoader.cs:60]: :+1:

**Review by captainsafia — APPROVED** [review]

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:242]: I've heard Larry talk about this so I guess it's the name of the archive file format. Apparently it's like a `tar` file.

**Review by SteveSandersonMS — APPROVED** [review]
Seems like this is mostly porting pre-existing changes so approving. Looks good anyway.

**Comment by pranavkm** [comment]
@lewing I'm seeing issues with non-en cultures. Here's the stack trace: 

```
_abort@dotnet.5.0.0-rc.1.20371.13.js:1
ves_icall_System_Environment_FailFast@00aba79e:1
ves_icall_System_Environment_FailFast_raw@00aba79e:1
do_icall@00aba79e:1
do_icall_wrapper@00aba79e:1
interp_exec_method@00aba79e:1
interp_runtime_invoke@00aba79e:1
mono_jit_runtime_invoke@00aba79e:1
do_runtime_invoke@00aba79e:1
mono_runtime_try_invoke@00aba79e:1
mono_runtime_try_invoke_handle@00aba79e:1
invoke_resolve_method@00aba79e:1
mono_alc_invoke_resolve_using_resolve_satellite@00aba79e:1
mono_alc_invoke_resolve_using_resolve_satellite_nofail@00aba79e:1
netcore_load_reference@00aba79e:1
mono_assembly_request_byname@00aba79e:1
ves_icall_System_Reflection_Assembly_InternalLoad@00aba79e:1
ves_icall_System_Reflection_Assembly_InternalLoad_raw@00aba79e:1
do_icall@00aba79e:1
do_icall_wrapper@00aba79e:1
interp_exec_method@00aba79e:1
interp_runtime_invoke@00aba79e:1
mono_jit_runtime_invoke@00aba79e:1
do_runtime_invoke@00aba79e:1
mono_runtime_try_invoke@00aba79e:1
mono_runtime_try_invoke_handle@00aba79e:1
invoke_resolve_method@00aba79e:1
mono_alc_invoke_resolve_using_resolve_satellite@00aba79e:1
```

Could you help?

**Comment by CoffeeFlux** [comment]
I'll take a look; getting this set up locally.

**Comment by CoffeeFlux** [comment]
Finally got a repro working, looking into this now.

**Comment by CoffeeFlux** [comment]
Building locally with the linker disabled and swapping out SPC for one built with https://github.com/dotnet/runtime/pull/40817 seems to fix this locally, so hopefully once that change flows down to this repo we'll be good to go. 😄

**Comment by benaadams** [comment]
ooof
image

**Comment by pranavkm** [comment]
@benaadams we're looking at sharding the icu file, which should make the default compressed size about half of the numbers being reported here. There's also an option to enable invariant mode if you do not require this data that should eliminate this file entirely.

**Pull request merged**